### PR TITLE
Clearify doSim and doSimUntilVoid use

### PR DESCRIPTION
**Pull request by andreasWallner**

### CODE REVIEW

**Comment by Dolu1990** [comment]
Hmmm 
"When in doubt about what to use, prefer ``doSim(...)``, for example"

personnaly i'm kind of always using doSimUntilVoid.
I mean, maybe instead we could provide a very minimal typical testbench for each ?

File: source/SpinalHDL/Simulation/bootstraps.rst [hunk 1]
```diff
@@ -79,18 +79,18 @@ This function will return a ``SimCompiled`` instance.
...
-   * - Syntax
-     - Description
-   * - ``doSim[(simName[, seed])]{dut => ...}``
-     - Run the simulation until the main thread is done (doesn't wait on forked threads) or until all threads are stuck
-   * - ``doSimUntilVoid[(simName[, seed])]{dut => ...}``
-     - Run the simulation until all threads are done or stuck
+``doSim[(simName[, seed])]{dut => /* main stimulus code */}``
+  Run the simulation until the main thread runs to completion and exits/returns.
+  It will detect and report an error if the simulation gets fully stuck. As long as
+  a e.g. a clock is running the simulation can continue forever, it is therefore recommended
```
Comment (numero-744) [inline source/SpinalHDL/Simulation/bootstraps.rst:85]:
```suggestion
  e.g. a clock is running the simulation can continue forever, it is therefore recommended
```

Repetition of a

File: source/SpinalHDL/Simulation/bootstraps.rst [hunk 2]
```diff
@@ -79,18 +79,18 @@ This function will return a ``SimCompiled`` instance.
...
+  to use ``simTimeout(cycles)`` to limit the possible runtime.
```
Comment (numero-744) [inline source/SpinalHDL/Simulation/bootstraps.rst:86]:
```suggestion
  to use ``SimTimeout(cycles)`` to limit the possible runtime.
```

I think it is `SimTimeout` not `simTimeout`.

File: source/SpinalHDL/Simulation/bootstraps.rst [hunk 3]
```diff
@@ -79,18 +79,18 @@ This function will return a ``SimCompiled`` instance.
...
 
+``doSimUntilVoid[(simName[, seed])]{dut => ...}``
+  Run the simulation until it is ended by calling either ``simSuccess()`` or ``simFailure()``.
+  The main stimulus thread can continue or exit early, as long as there are events to process,
```
Comment (numero-744) [inline source/SpinalHDL/Simulation/bootstraps.rst:90]: I am not sure I understand: does `simSuccess` stop simulation or not?

File: source/SpinalHDL/Simulation/bootstraps.rst [hunk 4]
```diff
@@ -103,6 +103,7 @@ For example :
      .workspacePath("~/tmp")
      .compile(new TopLevel)
      .doSim { dut =>
+       simTimeout(1000)
```
Comment (numero-744) [inline source/SpinalHDL/Simulation/bootstraps.rst:106]:
```suggestion
       SimTimeout(1000)
```

File: source/SpinalHDL/Simulation/bootstraps.rst (hunk 3 above)
Comment (andreasWallner) [reply source/SpinalHDL/Simulation/bootstraps.rst:90]: `simSuccess` will stop the simulation

**Comment by andreasWallner** [comment]
> personnaly i'm kind of always using doSimUntilVoid.

Interesting, I had never used it before quickly trying it when I wrote about it... (also: in the SpinalHDL repo there are 152 usages of doSim and 48 of doSimUntilVoid ;-))
I'll add a sample for `doSimUntilVoid` as well

**Comment by Dolu1990** [comment]
@andreasWallner 

May this example help : ? 

```scala
object SimMiaouDoc extends App {
  class TopLevel extends Component {
    val counter = out(Reg(UInt(8 bits)) init (0))
    counter := counter + 1
  }

  SimConfig.compile(new TopLevel).doSim { dut =>
    SimTimeout(1000)
    dut.clockDomain.forkStimulus(10)
    dut.clockDomain.waitSamplingWhere(dut.counter.toInt == 20)
  }

  SimConfig.compile(new TopLevel).doSimUntilVoid{ dut =>
    SimTimeout(1000)
    dut.clockDomain.forkStimulus(10)
    fork {
      dut.clockDomain.waitSamplingWhere(dut.counter.toInt == 20)
      simSuccess()
    }
  }
}
```

File: source/SpinalHDL/Simulation/bootstraps.rst (hunk 3 above)
Comment (numero-744) [reply source/SpinalHDL/Simulation/bootstraps.rst:90]: So I find this sentence being just after the end functions confusing "The main stimulus thread can continue or exit early".

What about telling that another way?

* the end of the block does not stop simulation
* simulation stops only when no more events to process, or called one of the two end functions

**Comment by numero-744** [comment]
With these examples, maybe a note explaining that with `doSim` the second example would not wait*?

*: Err… wait is not really visible, maybe adding a `println("done")` before `simSuccess()` and tell it does not print if using `doSim` instead of `doSimUntilVoid`?

**Comment by Dolu1990** [comment]
@numero-744 

> *: Err… wait is not really visible, maybe adding a println("done") before simSuccess() and tell it does not print if using doSim instead of doSimUntilVoid?

Ahhh sure ^^, why not :)

**Comment by Dolu1990** [comment]
@andreasWallner Would that be good for you ?

File: source/SpinalHDL/Simulation/bootstraps.rst [hunk 5]
```diff
@@ -79,18 +79,18 @@ This function will return a ``SimCompiled`` instance.
...
+``doSim[(simName[, seed])]{dut => /* main stimulus code */}``
+  Run the simulation until the main thread runs to completion and exits/returns.
+  It will detect and report an error if the simulation gets fully stuck. As long as
+  e.g. a clock is running the simulation can continue forever, it is therefore recommended
+  to use ``SimTimeout(cycles)`` to limit the possible runtime.
 
+``doSimUntilVoid[(simName[, seed])]{dut => ...}``
+  Run the simulation until it is ended by calling either ``simSuccess()`` or ``simFailure()``.
+  The main stimulus thread can continue or exit early, as long as there are events to process,
+  the simulation will continue. The simulation will report an error if it gets fully stuck.
```
Comment (numero-744) [inline source/SpinalHDL/Simulation/bootstraps.rst:91]: I think there is an issue with the code: wouldn't a function named `doSimUntilVoid` be expected to end successfully when void is reach / there are no events to process anymore?

Comment (Dolu1990) [reply source/SpinalHDL/Simulation/bootstraps.rst:91]: That was the initial intend, but it is kinda never used at all and not usefull, and instead provide some easy to produce false SimSuccess.
So we turned it into SimFailure if the simulation has nothing scheduled anymore.

**Comment by Dolu1990** [comment]
I added the example :)

File: source/SpinalHDL/Simulation/bootstraps.rst (hunk 5 above)
Comment (numero-744) [reply source/SpinalHDL/Simulation/bootstraps.rst:91]: So the name did match the behavior, but it does not anymore… So the name does not match the behavior!

File: source/SpinalHDL/Simulation/bootstraps.rst [hunk 6]
```diff
@@ -87,13 +87,22 @@ On this ``SimCompiled`` instance you can run your simulation with the following
...
+The following testbench template will use the following toplevel : 
 
 .. code-block:: scala
+    
+   class TopLevel extends Component {
+      val counter = out(Reg(UInt(8 bits)) init (0))
+      counter := counter + 1
+   }
+
+Here is a template with many simulation configuration :
```
Comment (numero-744) [inline source/SpinalHDL/Simulation/bootstraps.rst:102]: Typos

```
Here is a template with many simulation configurations:
```

File: source/SpinalHDL/Simulation/bootstraps.rst [hunk 7]
```diff
@@ -107,6 +116,45 @@ When in doubt about what to use, prefer ``doSim(...)``, for example :
        // Simulation code here
    }
 
+Here is a template where the simulation end by completting the simulation main thread execution :
```
Comment (numero-744) [inline source/SpinalHDL/Simulation/bootstraps.rst:119]:
```
Here is a template where the simulation ends by completing the simulation main thread execution:
```

File: source/SpinalHDL/Simulation/bootstraps.rst [hunk 8]
```diff
@@ -107,6 +116,45 @@ When in doubt about what to use, prefer ``doSim(...)``, for example :
...
+.. code-block:: scala
+
+    SimConfig.compile(new TopLevel).doSim { dut =>
+      SimTimeout(1000)
+      dut.clockDomain.forkStimulus(10)
+      dut.clockDomain.waitSamplingWhere(dut.counter.toInt == 20)
+      println("done")
+    }
+    
+Here is a template where the simulation end by explicitly calling a simSuccess() :
```
Comment (numero-744) [inline source/SpinalHDL/Simulation/bootstraps.rst:130]:
```
Here is a template where the simulation ends by explicitly calling `simSuccess()`:
```

File: source/SpinalHDL/Simulation/bootstraps.rst [hunk 9]
```diff
@@ -107,6 +116,45 @@ When in doubt about what to use, prefer ``doSim(...)``, for example :
...
+      SimTimeout(1000)
+      dut.clockDomain.forkStimulus(10)
+      fork {
+        dut.clockDomain.waitSamplingWhere(dut.counter.toInt == 20)
+        println("done")
+        simSuccess()
+      }
+    }
+
+Note is it equivalent to :
```
Comment (numero-744) [inline source/SpinalHDL/Simulation/bootstraps.rst:144]:
```
Note is it equivalent to:

```

Comment (numero-744) [reply source/SpinalHDL/Simulation/bootstraps.rst:144]: Missing empty line here provokes an error, which was reported but was ignored and the branch was merged. I think this merge was premature, and waiting for a review (even if you are the manager) would have avoided that.

See just above this title: https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Simulation/bootstraps.html#running-multiple-tests-on-the-same-hardware

**Comment by numero-744** [comment]
Oh I forgot to click "update" on one of my comments above, sorry ^^'

I had suggested this: https://github.com/andreasWallner/SpinalDoc-RTD/pull/2/files

Btw, I liked the example you used initially because it was extending `App`, which makes it easy to test quickly by copy-pasting it, but it it true that it does not let us add comments as text between lines.

File: source/SpinalHDL/Simulation/bootstraps.rst (hunk 6 above)
Comment (Dolu1990) [reply source/SpinalHDL/Simulation/bootstraps.rst:102]: Hmm i would say fondamentaly, doSimUntilVoid name do not specify anything about what hapen once void is here.

Typo fixed ^^

Comment (numero-744) [reply source/SpinalHDL/Simulation/bootstraps.rst:102]: > fondamentaly, `doSimUntilVoid` name do not specify anything about what hapen once void is here.

You got the point :smile:

**Comment by numero-744** [comment]
I understand

**Comment by andreasWallner** [comment]
Sorry for the silence, I was quite sick the last few days, I'll have a look later

**Comment by Dolu1990** [comment]
No worries ^^

**Comment by andreasWallner** [comment]
I had a quick look and just changed one of the sentences and fixed a missing newline that yielded a broken output. Looks OK from my point of view but: I'm not sure the examples make one appreciate the differences or advantages of the approaches - mostly because there is only a single line difference.
I was originally thinking to link to the examples that are already there in the documentation and are a good showcase, namely https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Simulation/examples/asynchronous.html and https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Simulation/examples/single_clock_fifo.html and if needed add a few words there why the approach was chosen. What do you guys think about that?

**Comment by andreasWallner** [comment]
Sorry - It took me until now until I saw that I was discussing on a merged PR...

**Comment by Dolu1990** [comment]
@andreasWallner  :D
So merged stuff doesn't mean can't be PR again for new changes :D
Nothing is definitive.

So, i think you are right, having link to the example may be a good thing instead of stuffing that section with too many.
What's about moving the added example into a new example section with more details ? Or yes, as you said, https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Simulation/examples/asynchronous.html and https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Simulation/examples/single_clock_fifo.html with some alternatives.

A PR can be open :)

**Pull request merged**